File: msg/COS_MutableStructure.pl
#!/usr/bin/perl -w
use strict;

use cosmic;

sub ShowMutableStructureLesson {
  my $txt = "";
  $txt .= "# OBJECT introduce simple mutable structures\n";
  $txt .= ShowLine(Op2("define",
		       "mutable-struct",
		       Proc("lst",
			    Op("let",
			       Paren(Paren("data",
					   Op2("map",
					       Proc("x", 
						    Op1("make-cell", 0)),
					       Ref("lst")))),
			       Proc("key",
				    Op("list-ref",
				       Ref("data"),
				       Op("list:find",
					  Ref("lst"),
					  Ref("key"))))))));

  

  $txt .= ShowLine(Op2("define",
		       "test-struct1",
		       Op1("mutable-struct",
			   ShowList("item1", "item2", "item3"))));
  $txt .= ShowLine(Op2("set!",
		       Op("test-struct1", "item1"),
		       15));
  $txt .= ShowLine(Op2("=",
		       Op1("get!",
			   Op("test-struct1", "item1")),
		       15));

  return $txt;
};


ShowLesson(ShowMutableStructureLesson());

File: msg/COS_MutableStructure.js
#!/usr/bin/env node

const cos = require('./cosmic');

cos.header('OBJECT', 'introduce simple mutable structures');

cos.add(`define mutable-struct | ? x:list |
  let ((cell:list | map (? x | make-cell 0) $x:list)) |
  ? x:find | list-ref $cell:list | list:find $x:list $x:find`);

cos.add(`define demo:mutable-struct | mutable-struct | vector x:1 x:2 x:3`);
cos.add(`set! (demo:mutable-struct x:1) 15`);
cos.add(`= 15 | get! | demo:mutable-struct x:1`);
